liquibase-core/src/main/java/liquibase/sqlgenerator/core/CreateProcedureGenerator.java

```diff
@@ -90,11 +93,33 @@ public ValidationErrors validate(CreateProcedureStatement statement, Database da
             procedureText = procedureText.replace("OR REPLACE", "");
             procedureText = procedureText.replaceAll("[\\s]{2,}", " ");
         }
-        sql.add(new UnparsedSql(procedureText, statement.getEndDelimiter()));
+        if (database instanceof MSSQLDatabase) {
+            CreateProcedureGenerator.getSqlStatementsWithSet(sql, procedureText, statement);
+        } else {
+            sql.add(new UnparsedSql(procedureText, statement.getEndDelimiter()));
+        }
         surroundWithSchemaSets(sql, statement.getSchemaName(), database);
         return sql.toArray(new Sql[sql.size()]);
     }
 
+    public static void getSqlStatementsWithSet(List<Sql> sql, String procedureText, CreateProcedureStatement statement) {
```

Comment:
the method name starts with `get...` but it returns nothing (void return type) and it seems confusing.

Also, it's not only checks given SQL content for `SET ... ON/OFF` it also adds Sql object to SQL list and I think that it should be moved to separate method.